### ISSUE TITLE
Changed file attributes of source code trigger a new build

### ISSUE THREAD

**Issue by Jac2NL** [heroku/terraform-provider-heroku#272]
### Heroku Provider Version
```
Terraform v0.12.26
+ provider.heroku v2.4.1
```

### Affected Resource(s)
- terraform-provider-heroku

### Terraform Configuration Files
```hcl
resource "heroku_build" "foobar" {
  app = "${heroku_app.foobar.id}"

  source = {
    # A local directory, changing its contents will
    # force a new build during `terraform apply`
    path = "../example-app"
  }
}
```

### Expected Behavior
According to the [docs](https://www.terraform.io/docs/providers/heroku/r/build.html#local-source): "When running terraform apply, if the contents (SHA256) of the source path changed since the last apply, then a new build will start".

### Actual Behavior
If the contents (SHA256) of the files in the path remain unchanged but attributes of those files have changed then a new build will be started. I did not expect this because the source code itself has not changed.

### Steps to Reproduce
1. `terraform apply`
2. `touch ../example-app/*`
3. `terraform apply`

### Important Factoids
I run Terraform from CircleCI (Docker image `hashicorp/terraform:light`). My app's source code will have different access/modification times every time CircleCI does a check out of the repository. This will force a new build because the source tar ball will have a different SHA256 checksum every time (the files themselves will have the same SHA256 checksums). And even if the tar balls were the same, gzipping them will create compressed tar balls that are different (gzip's last modified time will change).

I interpreted the documentation to mean that builds are triggered if the contents of files have changed. Mentioning the SHA256 checksum set me on the wrong foot because it implies only the contents of the files will be examined. If this behaviour is as expected then the documentation should be updated to say: "When running terraform apply, if anything in the source path changed since the last apply (file contents, timestamps, permissions or ownerships), then a new build will start".

IMHO changed file attributes should not trigger a rebuild. I worked around the problem by having CircleCI create a tar ball of the source code in the path. To make sure that the same SHA256 hash is generated if file contents have not changed I had to set all timestamps to a fixed value using `touch -t 197001010000` after checking out the code from the repo. I also had to add all files to the tar ball in sorted order. `generateSourceTarball` already does the latter but would it be an idea to have it ignore timestamps and other attributes too? 

You might also want to look into bzip2 for compressing the tar ball instead of gzip because bzip2 will have consistent hashes unlike gzip.

**Comment by mars** [comment]
Thanks for posting this issue @Jac2NL 👌

You're totally correct that [heroku_build source.path](https://www.terraform.io/docs/providers/heroku/r/build.html#local-source) will compare the SHA256s of two different tarball (tar+zip) archives, and it follows that metadata (file timestamps) contained in the archives will effect the comparison too.

We'll update the documentation to be clear about this. Setting all timestamps to the same value in every archive is a nice fix 👍 if you don't care about timestamps (some web apps might use timestamps in meaningful ways).

I'm pretty sure Heroku explicitly requires `.tar.gz` type build archives ([Heroku docs, Building and Releasing](https://devcenter.heroku.com/articles/build-and-release-using-the-api)), so bzip is not an option.

Maybe there are some specific files that are changing which are not the actual source, but other meta, IDE/tools, git/vcs? The `.git/` directory is not ignored for these. Nothing is. Do we need to make an ignore-list for local source builds? Use `.gitignore`?

**Comment by Jac2NL** [comment]
Hello Mars,

Thank you for your support! It appears that using the source.path option with a directory will always trigger a new build. Even if the contents and the metadata have not changed the gzip will make sure that a new checksum is generated and this will trigger a new build. If this is indeed the case then you might want to add a notice to the documentation.

My source path directory did not contain any VCS files. 

Have you considered adding an option to not gzip the tar ball? Uploading uncompressed tar balls is supported by Heroku. If the tar ball has not changed and is not compressed with gzip then this will not trigger a build.

**Comment by hugbubby** [comment]
> Hello Mars,
> 
> Thank you for your support! It appears that using the source.path option with a directory will always trigger a new build. Even if the contents and the metadata have not changed the gzip will make sure that a new checksum is generated and this will trigger a new build. If this is indeed the case then you might want to add a notice to the documentation.
> 
> My source path directory did not contain any VCS files.
> 
> Have you considered adding an option to not gzip the tar ball? Uploading uncompressed tar balls is supported by Heroku. If the tar ball has not changed and is not compressed with gzip then this will not trigger a build.

This is my experience and has been the behavior of this provider for the last several months I've been using it. Heroku always generates a new build, regardless of the contents or attributes of the source directory.

**Comment by hugbubby** [comment]
I use github actions to create a new build, and so the timestamps on the files in the directory, while unused by my program, _always_ trigger a new build. This delays terraform deploys that should take around 5s to 7min+

**Comment by robbiecloset** [comment]
>I worked around the problem by having CircleCI create a tar ball of the source code in the path. To make sure that the same SHA256 hash is generated if file contents have not changed I had to set all timestamps to a fixed value using touch -t 197001010000 after checking out the code from the repo. I also had to add all files to the tar ball in sorted order.

@Jac2NL any chance you should share a snippet of how you accomplished this? I am trying to do the same but it doesn't appear to be working. Thanks!

**Comment by Jac2NL** [comment]
Sure, I don't know if you need more CircleCI context but basically it runs this:

            echo "export TF_VAR_project_environment=\"<< parameters.my-env >>\"" >> $BASH_ENV
            echo "export TF_VAR_app_flavor=\"<< parameters.my-flavor >>\"" >> $BASH_ENV
            echo "export TF_VAR_app_team=\"<< parameters.my-team >>\"" >> $BASH_ENV
            echo "export TF_VAR_app_count=<< parameters.my-count >>" >> $BASH_ENV
            echo "export TF_VAR_app_name=<< parameters.my-workspace >>" >> $BASH_ENV
            source $BASH_ENV
            terraform init
            cd ˜/project
            find . -exec touch -t 202001010000.00 {} +
            find . -type f -print | grep -v -E "(^\.\/terraform|^\.\/\.git|^\.\/\.circleci)" | sort | tar -cf heroku-app.tar -T -
            terraform plan

The exact timestamp does not matter as long as you stick to a particular one. 
And the Terraform files checked out by CircleCI contain the following, which will be executed by the `terraform plan`:

            resource "heroku_app" "ci-demo" {
              name   = var.app_name
              region = "eu"
              organization {
                name     = var.app_team
                personal = var.app_team == "personal"
              }
            }

            resource "heroku_build" "ci-demo" {
              app = heroku_app.ci-demo.name

              source = {
                path = "../../heroku-app.tar"
              }
            }

            resource "heroku_formation" "ci-demo" {
              app        = heroku_app.ci-demo.name
              type       = "web"
              quantity   = var.app_count
              size       = var.app_flavor
              depends_on = [heroku_build.ci-demo]
            }

**Comment by mars** [comment]
This should be fixed by PR #319 which was released in [v4.6.0](https://github.com/heroku/terraform-provider-heroku/releases/tag/v4.6.0).
